fairroot/base/steer/FairRun.h

```diff
@@ -247,16 +249,19 @@ class FairRun : public TNamed
     /**
      * Call FillEventHeader on the source
      */
-    void FillEventHeader()
+    [[deprecated("Use GetSource()->FillEventHeader()")]] void FillEventHeader()
```

Comment:
If we recommend to unconditionally dereference the return value of `GetSource()`, we should also make sure it never returns a `nullptr`, right?

The C++ idiomatic way to communicate this would be to return a reference instead of a raw pointer, but FairRoot rarely uses references ..

Perhaps we should add new *safer-to-use* APIs to support call chains?
```cpp
FairSource& GetDataSource();
FairSource const& GetDataSource() const;
```

---

`GetEventHeader()`'s implementation currently guarantees not to return a `nullptr`, but I would still add a comment to the header file documenting this intended post condition that it does not get violated in the future by accident. Additionally, we should add unit tests.

Reply:
Yes, the common way for a getter function is to return a const ref. Returning a raw pointer requires the caller to constantly check whether it's a nullptr.

We could implement additional getters to return references. But for `GetSource()`, is it possible to do one of these?
1. Throw an exception `std::runtime_error()` if source is nullptr
2. Use return gsl::not_null<FairSource*>. This will call std::terminate automatically when fSource is a nullptr. By this way we don't need to change the API.

Reply:
I would love to fix some of the return types of existing member functions, but how can one do that gracefully in C++ (with a deprecation period)?

Reply:
Assuming we are not going to break API, `gsl::not_null` is pretty good shot for those member functions returning a pointer. 

For reference, I think we need two getters returning const ref and ref:
```cpp
auto Get_cref() -> const Type&;
auto Get_ref() -> Type&;
``` 
I wouldn't say the following is a good choice:
```cpp
auto Get() const -> const Type&;
auto Get() -> Type&;
```
because in many cases when the object is not a const, we still need to get a const ref to its member variable for the reading. Some classes like FairRun can never be a const. But get a const ref of, for example, event header for its reading is still necessary.